### PR TITLE
TINY-9242: Phase V => Adding scrolling support to Docking and Classic Mode Sticky Toolbars

### PR DESCRIPTION
**Pull request by ephox-mogran**
Related Ticket: TINY-9242

Description of Changes:
* TBA

Pre-checks:
* [N/A] Changelog entry added  - deferring for main merge
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [X] Milestone set
* [N/A] Docs ticket created (if applicable) - deferring

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/themes/silver/main/ts/Events.ts [hunk 1]
```diff
@@ -60,6 +62,27 @@ const setup = (editor: Editor, mothership: Gui.GuiSystem, uiMotherships: Gui.Gui
...
   };
 
+  const onElementScroll = DomEvent.capture(doc, 'scroll', (evt) => {
+    window.requestAnimationFrame(() => {
+      const c = editor.getContainer();
+      // Because this can fire before the editor is rendered, we need to stop that from happening.
+      // Some tests can create this situation, and then we get a Node name null or defined error.
+      if (c !== undefined && c !== null) {
+        // eslint-disable-next-line no-console
+        console.log('dispatching element scroll');
```
Comment (ephox-mogran) [inline modules/tinymce/src/themes/silver/main/ts/Events.ts:72]: - [ ] remove

File: modules/tinymce/src/themes/silver/main/ts/ui/dialog/WindowManager.ts [hunk 2]
```diff
@@ -46,7 +47,31 @@ const inlineAdditionalBehaviours = (editor: Editor, isStickyToolbar: boolean, is
           fadeOutClass: 'tox-dialog-dock-fadeout',
           transitionClass: 'tox-dialog-dock-transition'
         },
-        modes: [ 'top' ]
+        modes: [ 'top' ],
+        lazyViewport: (comp) => {
+          // If we don't have a special scrolling environment, then just use the default
+          // viewport of (window)
+          // TODO: Should we consult ui_of_tomorrow here also?
```
Comment (ephox-mogran) [inline modules/tinymce/src/themes/silver/main/ts/ui/dialog/WindowManager.ts:54]: - [ ] only execute when `ui_of_tomorrow: true`

File: modules/tinymce/src/themes/silver/main/ts/ui/header/StickyHeader.ts [hunk 3]
```diff
@@ -209,11 +210,36 @@ const getBehaviours = (editor: Editor, sharedBackstage: UiFactoryBackstageShared
...
+            const optScrollingContext = ScrollingContext.detect(comp.element);
+            return optScrollingContext.fold(
+              () => {
+                // Force the header to hide before it overflows outside the container
+                const boxHeight = box.height - headerHeight;
+                const topBound = box.y + (isDockedMode(comp, 'top') ? 0 : headerHeight);
+                return Boxes.bounds(box.x, topBound, box.width, boxHeight);
+              },
+              (scrollEnv) => {
+                // It might be docked at the window, or the scroller. Simplify for now.
```
Comment (ephox-mogran) [inline modules/tinymce/src/themes/silver/main/ts/ui/header/StickyHeader.ts:225]: - [ ] remove comment about simplify (it is out-dated)

File: modules/tinymce/src/themes/silver/main/ts/ui/dialog/WindowManager.ts (hunk 2 above)
Comment (ephox-mogran) [reply modules/tinymce/src/themes/silver/main/ts/ui/dialog/WindowManager.ts:54]: Actually, make a `ScrollingContext.detectWhenUiOfTomorrow` which takes `editor` as well and checks the option.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/scrolling/InlineDialogInScrollingContainerTest.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { describe } from '@ephox/bedrock-client';
+
+describe('browser.tinymce.themes.silver.editor.scrolling.InlineDialogInScrollingContainerTest', () => {
```
Comment (ephox-mogran) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/scrolling/InlineDialogInScrollingContainerTest.ts:3]: - [ ] delete file

File: modules/agar/src/test/ts/browser/assertions/ApproxComparisonsTest.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    });
+
+    it('wrong unit', () => {
+      assert.throws(
+        () => ApproxComparisons.measurement(10, 'px', 0).strAssert('Label', '10rem'),
+        '"10rem" did not have the correct unit. Expected: "px", but was: "rem"'
+      );
+    });
+
+    it('wrong unit, wrong margin', () => {
```
Comment (tiny-ben-tran) [inline modules/agar/src/test/ts/browser/assertions/ApproxComparisonsTest.ts:30]: This test doesn't have anything to do with margin. Should there be a separated test for `wrong margin`?

**Review by MitchCTiny — APPROVED** [review]

**Review by spocke — APPROVED** [review]

File: modules/agar/src/test/ts/browser/assertions/ApproxComparisonsTest.ts (hunk 5 above)
Comment (tiny-ben-tran) [reply modules/agar/src/test/ts/browser/assertions/ApproxComparisonsTest.ts:30]: Testing for the correct unit came first. Hence it didn't get to compare the margin

**Review by tiny-ben-tran — APPROVED** [review]

**Pull request merged**